### ISSUE TITLE
Automatic upload S3Object to bucket fails

### ISSUE THREAD

**Issue by end3r117** [awslabs/aws-mobile-appsync-sdk-ios#25]
Hey all,

This is a follow up to question I posted on the AWSAppSync forums and StackOverflow (links at bottom), which has some more detail on permissions and resolvers, but I'll summarize why I don't think that's an issue, and update the question here.

Basically, my app requires offline support, which AppSync provides almost perfectly, except for the file upload to an S3 bucket. I cannot get it to upload automatically. I intend to use User Pools, as that is my app's authentication method, but I have tried using IAM as well and cannot get the files to automatically upload. If I use S3TransferManager.default() and manually make the upload request using either authentication method, it succeeds every time, so I believe my permissions are correct. The S3Object is also added as an item attribute in DynamoDB as such:
`"pdf": "{\"s3\":{\"key\":\"example.pdf\",\"bucket\":\"#####\",\"region\":\"us-west-2\"}}"`

I set breakpoints in your AppSync code to check if the call to upload ever happens, and it doesn't seem to. This seems to be because, in line 508 in AWSAppSyncClient.swift (AppSync 2.6.15 using codegen w/ s3 wrapper argument), the check

`if self.mutationRecord.type == .graphQLMutationWithS3Object { ...`

always fails, so

`appSyncClient.performMutationWithS3Object(...)`

is not called, but instead

`appSyncClient.send(...)`

is called.

After a query, I am also able to successfully use
```
guard let s3Object = exampleThing?.pdf?.fragments.s3Object else { return }
appSyncClient?.s3ObjectManager?.download(s3Object: s3Object, toURL: fileURL!, completion: { (success, error) in
  ...
}

```

to download example.pdf (if I upload it to the S3 bucket myself, of course).

I followed your tutorials and created the S3Object type and S3ObjectInput in the schema, connected the appropriate resolvers, and it would seem set the right permissions. Do you guys know what I'm doing wrong here? Why is mutationRecord.type == .graphQLMutationWithS3Object not returning true or, more generally, why is the automatic upload to S3 not happening?

Currently, I've just created my own offline solution that uploads the files I need to S3 when Reachability goes from .none to either wifi or cellular, but it would be ideal if the upload was handled by AppSync at the same time it does the offline mutation.

Any ideas? Thanks!

PS: Also, and this could just be a breakpoint thing, mutationRecord.type shows up as `<invalid>` to my debugging window, while the other variables are 'some', etc. And mutationRecord.s3ObjectInput is 'none' (this breakpoint was set during a CreateThingMutation, after checking mutationRecord.type fails, inside the 'else' statement that calls appSyncClient.send(...)).

AWS AppSync Forum question:
https://forums.aws.amazon.com/message.jspa?messageID=847113
S.O. question:
https://stackoverflow.com/questions/50050340/aws-appsync-unable-to-upload-s3object-to-bucket-automatically-with-cognito-use

**Comment by rohandubal** [comment]
Hi @end3r117 

Thanks for reporting this! I will investigate this issue and post an update here.

Thanks,
Rohan

**Comment by rohandubal** [comment]
Hi @end3r117 

Does your mutation with complex objects have the s3 file at the root level? Currently the SDK checks for the file to be at the root level for the upload to happen. 

Have you defined the local schema to use the mutation to use `S3ObjectInput`? For example, the `.graphql` file in the local iOS dev environment should have a mutation which looks like this:

```
mutation CreatePost($postInput:PostInput!, $image:S3ObjectInput!){
    createPost(postMetadata:$postInput, image:$image){
        id
    }
}
```

The decision of whether the mutation is of type `graphQLMutationWithS3Object` is made early over here:
https://github.com/awslabs/aws-mobile-appsync-sdk-ios/blob/master/AWSAppSyncClient/AWSAppSyncClient.swift#L470

This record is then persisted and retrieved. If the record is not being classified as `graphQLMutationWithS3Object` it is most likely that the input is not in the expected format; which is validated by this method: https://github.com/awslabs/aws-mobile-appsync-sdk-ios/blob/6fed18399851c7aa6951fef09ceb976854566606/AWSAppSyncClient/AWSAppSyncClient.swift#L487

Also have you followed the documentation posted here:
https://awslabs.github.io/aws-mobile-appsync-sdk-ios/#complex-objects 

The doc above contains step by step info on how to add complex objects to your schema.

Thanks,
Rohan

**Comment by end3r117** [comment]
Hey, Rohan. I am out of town for the next week and a half but I will try your suggestions, check file locations, and reply with an update ASAP. Please do not close this issue due to my temporary idleness!

Thanks for the help

**Comment by rohandubal** [comment]
We also have updated instructions which walk through the schema setup and client code here: https://awslabs.github.io/aws-mobile-appsync-sdk-ios/#complex-objects which you can try. Closing due to lack of activity.

**Comment by ajzhou** [comment]
I am having the same issue here. I followed the updated instructions but still have no success with uploading to S3 automatically through AppSync. I checked to upload the same file URI with TransferUtility and everything works.

**Comment by palpatim** [comment]
Commenting on this old issue to note that this issue is fixed as of release 2.9.0.
